Add `DEPLOYING.md` and `UPGRADING.md` documents

bickelj opened the pull request:
The documentation in this repo so far has descriptions of the elements
in a potential deployment pipeline but lacked a step-by-step guide to
deploy PDC in the actual instance of the PDC's pipeline. This commit
resolves the gap by adding more concrete guides in `DEPLOYING.md` and
`UPGRADING.md`. "Deploying" here means "how to deploy to production"
and "upgrading" here means "how to update a service-level dependency
such as postgresql in production."

Issue https://code.librehq.com/ots/infra/admin-tasks/-/issues/63
Issue https://code.librehq.com/ots/infra/admin-tasks/-/issues/64

kfogel commented on README.md:
```diff
@@ -9,6 +9,9 @@ component repositories because they operate above a source-code level, are
 service's level.
 
+This README document describes components of a deployment pipeline. For specific
+steps to deploy using the existing main deployment pipeline, see [deploying.md].
```
Did you test that all the Markdown renders as expected?  I don't think that "[deploying.md]" does what you think it does.  (You can render Markdown->HTML locally to test -- [this function](https://viewvc.red-bean.com/kfogel/trunk/.emacs?revision=6560&view=markup#l6828) in my `.emacs` is how I do it from Emacs, but you could just swipe the `pandoc -f gfm ...` command from there and use it however you want, of course.)

kfogel replied:
Also, you can just say "This README" instead of "This README document" :-).  (It's like saying "the `foo` variable has the value 3" instead of just "`foo` has the value 3".)

bickelj replied:
> Did you test that all the Markdown renders as expected? I don't think that "[deploying.md]" does what you think it does.

D'oh, nope.

> (You can render Markdown->HTML locally to test -- [this function](https://viewvc.red-bean.com/kfogel/trunk/.emacs?revision=6560&view=markup#l6828) in my `.emacs` is how I do it from Emacs, but you could just swipe the `pandoc -f gfm ...` command from there and use it however you want, of course.)

This is now in my `init.el`. After installing pandoc and markdown it just works. I also found a system-level package of ELPA's markdown mode and that helps too.

bickelj replied:
Thanks to your Emacs function, it was easy to test the links! Thank you!

kfogel replied:
Ah, great! You're welcome.  I use that function all the time.

kfogel commented on deploying.md:
```diff
@@ -0,0 +1,113 @@
+# Deploying Philanthropy Data Commons software
```
Typically, top-level documentation files, like `README.md` and `CONTRIBUTING.md` and such, have names in all caps.  I don't know why, but that's the convention (I suppose I could guess at reasons why that convention has evolved -- I do think it's a good one).  By that convention, this should be `DEPLOYMENT.md`.

kfogel commented on deploying.md:
```diff
@@ -0,0 +1,113 @@
+https://api.philanthropydatacommons.org and
+https://auth.philanthropydatacommons.org.
+
+The front-end at https://pilot.philanthropydatacommons.org has its own process.
```
Where is that process documented?  We might as well point to it from here.

bickelj replied:
:thinking: :construction:

bickelj replied:
I put a best guess link in for now, but I await a pointer from @reefdog or @jasonaowen regarding the deployment docs there.

jasonaowen replied:
Rather than specifically link to a Netlify page, I'd suggest just linking to the root of the data-viewer repo. That way, if we choose to switch away from Netlify, there's no risk of forgetting to update this repo:

```md
The [data-viewer](https://github.com/PhilanthropyDataCommons/data-viewer/) at
https://pilot.philanthropydatacommons.org has its own separate process.
```

kfogel commented on deploying.md:
```diff
@@ -0,0 +1,113 @@
+
+The front-end at https://pilot.philanthropydatacommons.org has its own process.
+
+What gets deployed? A [docker compose yaml file][1].
```
I didn't understand this sentence.  What gets deployed is the service itself, no?  A Docker compose YAML file is just part of the means by which that happens -- that file is a specification for the deployment, but that file is not itself being deployed.

bickelj replied:
In a literal sense, the docker compose yaml file gets copied to the machine and could be thought of as "deployed", but you're right: the yaml file is a means to an end. The goal is to deploy the service and all its dependencies.

The reason to say that it's the yaml file that gets deployed is because that yaml file is versioned and that version of the yaml file is what you specify to deploy. This is in contrast to the version of the service image, which is one level removed, modeled as a dependency of the yaml file.

I'll think of a better way to phrase it.

bickelj replied:
Replaced with this:
> What gets deployed? Everything specified in a [docker compose yaml file][2].
> This includes the PDC service ("web") and other services upon which it depends.

kfogel commented on deploying.md:
```diff
@@ -0,0 +1,113 @@
+[2]: https://github.com/PhilanthropyDataCommons/deploy/blob/main/deploy.sh
+[3]: https://github.com/PhilanthropyDataCommons/deploy/blob/e0d24b8a19ee563938f8211bca0485a97ff1a6e9/deploy.sh#L90
+
+## The common case: deploy a newly built version of the PDC service
```
This is probably not the common case, in the long run.  The most common case would probably be a local deployment for development purposes.  Actual production (or staging) rollouts should be fairly uncommon compared to local deployments.

bickelj replied:
Until you mentioned it, I didn't even think of a local deployment as a deployment, but rather something like "setting up a development environment." Local development would be covered by https://github.com/PhilanthropyDataCommons/service#development. I suppose it should be made more clear that this document is focused on production deployments and add a link to that local development.

bickelj replied:
I can remove the judgment, though, to your point. It's unneeded.

bickelj replied:
Simplified to this:
> ## To deploy a newly built version of the PDC service to production

kfogel commented on deploying.md:
```diff
@@ -0,0 +1,113 @@
+
+### Optionally review this background information
+
+On a merge to the main branch in the [service repository][4], tests run and a
```
"When a branch is merged into `main` in the service repository, GitHub Actions will automatically run the test suite and then, assuming the tests succeed, build a new docker image containing the code and all its dependencies.  That image will be tagged with a tag like `give-an-actual-example-tag-here`. The tag is on the binary image, not on the source code, but the tag's text is based on the git commit, so that we can always trace the image back to the source code whence it came."

That wording is just a suggestion.  The main thing is to steer clear of subjectless voice as much as possible in this kind of documentation.  Always say exactly what actor performs what actions, with what results, in what order.

In the rest of the document, you seem to mostly use the latter style, and it's good IMHO.

kfogel commented on upgrading.md:
```diff
@@ -0,0 +1,170 @@
+# Upgrading Philanthropy Data Commons dependencies
```
Similarly, upper case filename before the extension here.  (Alternatively, you could move documents to a `docs/` subdir.)

kfogel commented on upgrading.md:
```diff
@@ -0,0 +1,170 @@
+## Upgrade postgresql (minor version)
+
+Caution: do not use "latest". That won't work the way we want. Do not use a
+non-specific version such as 14-debian-11 or 14.8.0. Use only specific versions.
```
How is "14.8.0" not a specific version?

bickelj replied:
There may be multiple images containing a 14.8.0 version of postgresql built and pushed to docker hub. So while it appears specific, and is a specific version of postgresql, we are working with particular built images and can specify exactly that image. I'll update the text to indicate "less specific" versus "more specific."

bickelj replied:
Changed it to this:
> Also do not use a
> less-specific version such as 14-debian-11 or 14.8.0 but only the most specific
> image versions such as 14.8.0-debian-11-r14.

kfogel commented on upgrading.md:
```diff
@@ -0,0 +1,170 @@
+  - `git add compose.yml`
+  - `git commit` (Use a message indicating the new version)
+  - `git push origin upgrade-to-postgresql-image-[version]`
+8. Create a pull request to merge that branch (follow the link git prints).
```
I think you don't need to say "(follow the link git prints)" here.

In general, you can make some assumptions about the audience of a document like this -- you have to, really.  One of those assumptions is that the person has basic familiarity with using GitHub (or other GitHub-like forge platforms like GitLab) as a developer.  I think that's a safe assumption.  We don't need to teach them how PRs work.

kfogel commented on upgrading.md:
```diff
@@ -0,0 +1,170 @@
+  - `git commit` (Use a message indicating the new version)
+  - `git push origin upgrade-to-postgresql-image-[version]`
+8. Create a pull request to merge that branch (follow the link git prints).
+9. Ask someone to review the pull request. If approved, merge it to main. Or if
```
Is this "wait one business day" for review a team principle?  I hadn't seen it before, and wondered where it came from.

bickelj replied:
We haven't been practicing it as much recently, but yes: https://chat.opentechstrategies.com/#narrow/stream/66-PDC-.28internal.29/topic/code.20reviews/near/147746. I should add the "give an hour or two last notice" part too.

kfogel commented on upgrading.md:
```diff
@@ -0,0 +1,170 @@
+The tag will be the date in UTC of the commit (in main) followed by the first
+seven digits of the git commit hash. For example, if the commit was like this:
+`e0d24b8a19ee563938f8211bca0485a97ff1a6e9...Tue Jun 13 19:00:38 2023 +0000`, the
+tag would be `20230613-e0d24b8`.
```
Nice example!

kfogel reviewed (commented):
I think one of the devs should review this, in addition to me -- they're the real audience, and are more qualified than I am to comment on much of it.

bickelj commented:
I just used the guide to deploy. It took 6 minutes and 24 seconds, give or take. The links seemed to work.

bickelj commented on DEPLOYING.md:
```diff
@@ -0,0 +1,123 @@
+
+ - `ssh $OTS_USERNAME@api.philanthropydatacommons.org`
+ - `sudo su - build`
+ - `echo "20230601-4a43cf2" > deployment/tag_to_deploy`
```
While trying it out, I realized that an actual version here might be counterproductive.

jasonaowen commented on DEPLOYING.md:
```diff
@@ -0,0 +1,126 @@
+How do they get deployed? A [bash script running under cron][3] gets the new
+compose file, pulls new components (images) using the new compose file, stops
+the old services (containers) using the old compose file, and starts the new
+services (containers) using the new compose file.
```
This is an intricate enough sentence that it would probably benefit from splitting out into bullet points:

```suggestion
The docker services are deployed by [bash script running under cron][3], which
- gets the new compose file
- pulls new components (images) using the new compose file
- stops the old services (containers) using the old compose file
- and starts the new services (containers) using the new compose file.
```

jasonaowen commented on DEPLOYING.md:
```diff
@@ -0,0 +1,126 @@
+
+### Tell the deploy script the version of the compose file to deploy
+
+These are the commands that actually deploy the software.
+
+ - `ssh $OTS_USERNAME@api.philanthropydatacommons.org`
+ - `sudo su - build`
+
+In the next command, replace VERSION with the compose file version:
+
+ - `echo "VERSION" > deployment/tag_to_deploy`
+ - `exit`
```
If you need an interactive shell, use `sudo -i` rather than `sudo su -`; in this case, to become the `build` user, use `sudo -iu build`. But, do you need an interactive shell? I'd suggest `echo "VERSION" | sudo -u build tee ~build/deployment/tag_to_deploy` or `sudoedit -u build ~build/deployment/tag_to_deploy`.

Stepping back from the documentation level to discuss the process itself, I feel like you ought not need to use `sudo` to deploy. Should your user be in the `build` group, via something like `sudo adduser $OTS_USERNAME build`?

bickelj replied:
I see the difference (with `sudo -iu build`) is more direct and results in one fewer shells. Nice.

For now I'm doing the `echo` suggestion but agree that the users who deploy should be in the `build` group and then they can do it exactly as the automated deployment does with a local echo to file followed by `scp...`.

jasonaowen commented on DEPLOYING.md:
```diff
@@ -0,0 +1,126 @@
+
+(Continuing the above session at `api.philanthropydatacommons.org` as
+`$OTS_USERNAME`):
+ - `sudo docker ps`
```
Likewise, you should be able to avoid docker commands requiring `sudo` by adding yourself to the `docker` group: `sudo adduser $OTS_USERNAME docker`.

bickelj replied:
Yes, but perhaps there is a better audit trail with the `sudo` command, right?

jasonaowen replied:
Perhaps so!

jasonaowen commented on DEPLOYING.md:
```diff
@@ -0,0 +1,126 @@
+(node) are running.
+3. Check the [auth service in the test environment][13]. If a JSON doc shows up,
+this means the reverse-proxy container (nginx), auth container (keycloak), and
+db container (postgresql) are running.n
```
Typo:

```suggestion
db container (postgresql) are running.
```

jasonaowen commented on DEPLOYING.md:
```diff
@@ -0,0 +1,126 @@
+machines were already set up with users, directories, docker, certbot, scripts,
+environment variables, additional keycloak jar files, and so forth.
+
+### Optionally review this background information
```
This section header can be simplified:

```suggestion
### Background
```

jasonaowen commented on DEPLOYING.md:
I'm not sure where exactly to anchor this conversation, so I'll just comment on the file.

I haven't gone through the deployment process, so it took me a while to understand this, and I'd appreciate being double-checked on it: the entire deploy-to-prod process is "edit a file and wait"? Is that right? If so, that's worth summarizing somewhere near the top, because it was not clear to me on the first few read-throughs.

bickelj replied:
Yes, or I would say "write a file and wait." That's assuming you've already set up the deployment pipeline.

I can't be sure if we mean the same thing by "deployment" so does that square with your understanding of "deployment"? Or would you include setting up that pipeline as "deployment"?

I added a line to `DEPLOYING.md` so this should be good enough for now.

jasonaowen replied:
> I added a line to `DEPLOYING.md` so this should be good enough for now.

Yes, thank you!

jasonaowen commented on DEPLOYING.md:
```diff
@@ -0,0 +1,126 @@
+the old services (containers) using the old compose file, and starts the new
+services (containers) using the new compose file.
+
+How does the deploy script know how what to deploy? It [looks inside a file][4].
```
I find the Q&A format used in this section makes it harder for me to understand this information; instead, I'd expect a more declarative tone, like "The deploy script looks inside a file to determine what docker compose script to use."

I'm also not clear from looking at the linked line what file is used:

https://github.com/PhilanthropyDataCommons/deploy/blob/e0d24b8a19ee563938f8211bca0485a97ff1a6e9/deploy.sh#L90

https://github.com/PhilanthropyDataCommons/deploy/blob/96df6a36c0a8ab9734d9681482d3850f466eb92f/deploy.sh#L68-L69

What file is actually used in production? From elsewhere in this doc, it appears to be `~build/deployment/tag_to_deploy`; is that right?

bickelj replied:
I took the "what file" comment as a suggestion and I added the value of `TAG_FILE`.

As for Q&A I am still in favor of the Q&A but obviously this document is not for me. It's not necessarily for you exclusively either, so I am curious what other folks think. As for why I care about the Q&A, I think the questions are relevant to all deployment pipelines, and the answers here are specific to this deployment pipeline, and the answers here may even change for this pipeline. But the format wouldn't change just because the implementation or answers changed.

@smpsnr or @kfogel? Q&A or simple declarations or something else?

kfogel replied:
@bickelj I understand the arguments for Q&A format, and they are reasonable, but FWIW I find that in practice it rarely helps me understand a document better.  My vote would be for straightforward declarative flow.  This isn't a hard and fast principle: occasionally one can insert a leading question when it clarifies to do so.

bickelj replied:
I removed the questions.

jasonaowen commented on UPGRADING.md:
```diff
@@ -0,0 +1,177 @@
+https://auth.philanthropydatacommons.org.
+
+The front-end at https://pilot.philanthropydatacommons.org has [its own separate
+process][1].
```
Does the data-viewer need to be mentioned at all here? I don't think it has a meaningful analogue to the types of dependencies this document describes.

```suggestion
```

jasonaowen commented on UPGRADING.md:
```diff
@@ -0,0 +1,177 @@
+on the host, such as cert-bot.
+
+See the summary in [DEPLOYING][3] regarding how software gets deployed. The same
+explanation regarding [compose.yml][4] applies.
```
Which `compose.yml` explanation are you referring to? The one in `README.md`?

jasonaowen commented on UPGRADING.md:
```diff
@@ -0,0 +1,177 @@
+
+## Upgrade postgresql (minor version)
+
+Caution: do not use "latest". That won't work the way we want. Also do not use a
```
Please describe why using latest "won't work the way we want" - what specific problem will using `latest` lead to, and why is that bad? I think I agree with this decision, but it's hard to say without an explanation of the reasoning that lead to it.

jasonaowen commented on UPGRADING.md:
```diff
@@ -0,0 +1,177 @@
+less-specific version such as 14-debian-11 or 14.8.0 but only the most specific
+image versions such as 14.8.0-debian-11-r14. The purpose of using bitnami images is
+reproducibility of issues by virtue of [reproducibility of images via immutable
+tags][5].
```
This feels like the start of an argument in favor of the given approach, but the given link does not contain the word "immutable". Is that the link you intended to use here? Can you say more?

jasonaowen commented on UPGRADING.md:
```diff
@@ -0,0 +1,177 @@
+line in a file and tagging it. There are multiple ways to do it. See [the
+official postgresql docs][14].
+
+[14]: https://www.postgresql.org/docs/15/upgrading.html
```
This is probably a case where we do not want a specific version - which is an irony I appreciate, in this doc :laughing: 

```suggestion
[14]: https://www.postgresql.org/docs/current/upgrading.html
```

bickelj replied:
I went back and forth on this, actually. On the one hand, we don't want a stale link. On the other hand, we want a reliable link with predictable content. I erred on the side of predictable. If the 16 and 17 docs are basically the same as the 15 docs, there is no loss in keeping the 15 link. If the 16 and 17 docs change quite a bit, then the link may not even work. In any case, it is good to know that at the time of writing 15 was the current version.

jasonaowen replied:
The PostgreSQL community knows how to run a stable website :slightly_smiling_face: you may consider the `current` links to be reliable.

My concern is that upgrading to 15 will need to involve updating this documentation, which is more work and has a risk of being forgotten. The same applies for Keycloak, but there you have more detailed and specific upgrade concerns, so we can leave this as-is for consistency.

jasonaowen commented on UPGRADING.md:
```diff
@@ -0,0 +1,177 @@
+
+## Upgrade postgresql (major version)
+
+A major postgresql version upgrade is quite a bit more involved than changing a
```
Please capitalize PostgreSQL [as they do](https://www.postgresql.org/about/) throughout this document when referring to the project.

```suggestion
A major PostgreSQL version upgrade is quite a bit more involved than changing a
```

jasonaowen commented on UPGRADING.md:
```diff
@@ -0,0 +1,177 @@
+order to perform a migration of the old data format to the new data format.
+
+To provide access to the binaries of the old postgresql image in the new image,
+copy them from the image to a new directory and mount it in the new image.
```
This seems strongly at odds with the Docker way of running a database - @bickelj, as someone who is more familiar with Docker than I am, would you agree with that assessment? If so, is it worth including this section at all?

bickelj replied:
I can't say it's strongly at odds until I try an alternative. I agree it is convoluted on the face of it. But it was one way to get it to work. If another way is superior, sure, we can remove the inferior once the superior details are fleshed out.

bickelj replied:
Having a list of commands that work to upgrade Postgres was one of the requests that led to this documentation. So I would rather not omit something that seemed to work fine without an alternative list of commands.

jasonaowen replied:
> I can't say it's strongly at odds [with the Docker way of running a database] until I try an alternative.

I just meant the amount of manual steps and local modifications needed seems like you're going against the grain of how Docker wants you to work. Are such steps typical in your experience of using Docker?

> Having a list of commands that work to upgrade Postgres was one of the requests that led to this documentation. So I would rather not omit something that seemed to work fine without an alternative list of commands.

Ah, I missed that request; that's helpful context. I'm glad you were able to get it to work, but the number of cautioning statements leaves me, well, cautious!

Perhaps it's worth splitting "what we did last time" out into the server updates log in Subversion? I agree that it's useful to have this list of commands, but putting them here implies a level of endorsement.

bickelj replied:
I would add it to svn docs if we had done it that way on the server. I've only tried it locally and took some notes. That's where these came from. We have yet to do a major version upgrade.

bickelj replied:
I added some notes to each section and put the presumed best section on top. I hope this resolves it.

jasonaowen commented on UPGRADING.md:
```diff
@@ -0,0 +1,177 @@
+
+See [the postgresql docs][17].
+
+[17]: https://www.postgresql.org/docs/15/upgrading.html#UPGRADING-VIA-REPLICATION
```
Do you plan on upgrading via replication? If not, let's drop this section!

bickelj replied:
This looks like the least-downtime method. Perhaps once we land on a primary recommended method (which I assume will be `pg_dumpall` and `pg_restore`) then the alternatives can be briefly mentioned in a single sentence or something like that.

Do you think I should try the `pg_dumpall` and `pg_restore` before we merge these documents?

jasonaowen replied:
> Do you think I should try the `pg_dumpall` and `pg_restore` before we merge these documents?

No, that's not necessary!

jasonaowen commented on UPGRADING.md:
```diff
@@ -0,0 +1,177 @@
+
+[15]: https://www.postgresql.org/docs/15/pgupgrade.html
+
+### The `pg_dumpall` and `pg_restore` way
+
+See [the postgresql docs][16].
+
+The steps should be to run a second postgresql container with the new version of
+postgresql and connect from it to the first, already-running postgresql
+container.
```
I would have expected this to be the approach we would use with Docker, so I am surprised to see how brief it is compared to the previous section.

I'd also summarize the steps as "run the new version of `pg_dumpall` against the old instance, and then restore into the new instance" - the two instances do not need to be running at the same time.

bickelj replied:
The reason it is brief is I haven't tried it this way.

bickelj replied:
I'm deleting the paragraph until I try it. My summary does not say that a new database instance has to be running, only the container with the new version of PostgreSQL. Your summary also is confusing because if `pg_dumpall` needs to connect, then the old instance needs to be running.

jasonaowen replied:
The old instance needs to be running to take a backup of it, yes. The backup should be taken with the newer version of `pg_dumpall`. The new cluster does not need to be running to use `pg_dumpall`, however, and the old cluster does not need to be running to use `psql` or `pg_restore` - the backup and restore may be separated in time.

bickelj replied:
Does the deletion resolve it for now?

jasonaowen commented on UPGRADING.md:
```diff
@@ -0,0 +1,177 @@
+8. Create a pull request to merge that branch.
+9. Ask someone to review the pull request. If approved, merge it to main. Or if
+no one reviews it by the same time on the next business day, give a last-hour
+notice for review and then merge it to main on approval or when no one reviews.
```
To echo @kfogel's [earlier](https://github.com/PhilanthropyDataCommons/deploy/pull/88#discussion_r1235673671) [comments](https://github.com/PhilanthropyDataCommons/deploy/pull/88#discussion_r1235676173), I don't think you need to spell out the entire pull request process here. These can be collapsed down to a single step, I think:

```suggestion
7. Create a pull request with your changes.
```

jasonaowen commented on DEPLOYING.md:
```diff
@@ -0,0 +1,126 @@
+What gets deployed? Everything specified in a [docker compose yaml file][2].
+This includes the PDC service ("web") and other services upon which it depends.
+
+How do they get deployed? A [bash script running under cron][3] gets the new
```
Is the crontab committed anywhere? I don't see it listed in this repo, and in particular the word "cron" does not appear in the given link.

bickelj replied:
Nope.

jasonaowen replied:
Okay. It should be. Do you want to do that as part of this PR, or in a separate issue?

bickelj replied:
I can add it here.

bickelj replied:
Added `deploy.crontab`

jasonaowen commented on UPGRADING.md:
```diff
@@ -0,0 +1,177 @@
+[10]: https://hub.docker.com/r/bitnami/postgresql/tags
+[11]: https://github.com/PhilanthropyDataCommons/deploy/issues/31
+
+## Upgrade keycloak (minor version)
```
Please capitalize Keycloak as they do throughout this document when referring to the project:

```suggestion
## Upgrade Keycloak (minor version)
```

bickelj replied:
For consistency, I have re-capitalized a few other names. The result is really ugly. But as you wish!

jasonaowen replied:
I'm not sure I understand the objection. What is an example of prose containing names that you find to be ugly when those names are capitalized?

jasonaowen replied:
To justify my position a bit, my stance is based on respect. We are fortunate to be able to rely on these projects, and the hard work so many people have contributed to them. The least we can do is [call them by their preferred names](https://www.youtube.com/watch?v=WssBJeExiOM)!

jasonaowen replied:
> For consistency, I have re-capitalized a few other names. The result is really ugly.

Aha, I believe you might be referring to replacing "nginx" with "NGINX Open Source" - is that right? If so, I agree that the latter is more awkward.

I actually had looked at that section, and looked up how the nginx project refers to itself. On their [about page](https://nginx.org/en/) they use all-lowercase "nginx", and the [nginx Wikipedia page](https://en.wikipedia.org/wiki/Nginx) says "stylized as **NGINX** or **nginx**". So, for that project, I'd suggest the lowercase styling.

kfogel replied:
I think I might have introduced "NginX" or something like that, here or in a nearby project.  That was based on a wrong memory.  My intention was to call the project by the name the project itself uses -- if I implemented that intention in a buggy way, it's just a bug to fix :-).

In general I agree with @jasonaowen here: let's write projects' names the way the projects themselves do.  It is both a sign of respect and helps train the eye to recognize commonly-encountered proper nouns quickly (which aids accuracy & speed in reading documentation -- that is to say, it reduces the error rate at any given reading speed).

bickelj replied:
Over-capitalization of tool names is annoying. People assume that every software tool is an abbreviation for something so you end up with folks referring to JAVA or NODE or whatever. When trying to read a sentence that refers to tools (not people, mind you), I find it distracting to have stuff like PostgreSQL and Node.js when postgres or node would work just as well. The binaries we run on the system are all lower case, as well. I suppose y'all's argument is that in formal documentation we should use formal names, not that in general we must use PostgreSQL and Node.js, right?

bickelj replied:
https://www.nginx.com/ always uses "NGINX Open Source" and I supposed they hold the trademark.

bickelj replied:
I'll revert to `nginx`, but I guess I will rely on review to figure out where the line is between the common names and the legal names.

reefdog replied:
> I suppose y'all's argument is that in formal documentation we should use formal names, not that in general we must use PostgreSQL and Node.js, right?

Not sure if you meant "formal documentation" to be a limiting statement, but I'd expand this to all uses of the project name. Anywhere we mention it, stylize it as they do. (This of course doesn't refer to the binary references, but for example it _does_ include mentioning the project in code comments.) When there are multiple canonical options (like, it seems NGINX/nginx), let's just be consistent. Among the other good reasons provided here, consistency aids searching for references.

bickelj replied:
@reefdog I'm hoping there's a limit somewhere, yes. For example in a chat during a call about a production outage, I hope no one stops to say "you said postgres when you really meant PostgreSQL." Or perhaps an in-line comment in a source file.

But if I'm really alone on the over-capitalization as a nuisance view the simpler rule is to learn the tool name and always write it with the legal name and I can start doing that. I don't think I will say aloud "post gress queue ell" though.

reefdog replied:
@bickelj I will happily join you in glaring and shaking a fist at anyone who insists we say "post gress queue ell" verbally. 😆

jasonaowen commented on DEPLOYING.md:
```diff
@@ -0,0 +1,126 @@
+the recently built and tested version of the PDC service to the production
+environment.
+
+[5]: https://github.com/PhilanthropyDataCommons/service
+[6]: https://github.com/PhilanthropyDataCommons/service/tree/main/.github/workflows
+[7]: https://github.com/PhilanthropyDataCommons/service/pkgs/container/service
+[8]: https://github.com/PhilanthropyDataCommons/service/blob/cd2e2a4c39d87479f7b145a9163e155ede7676f2/.github/workflows/build.yml#L46
+[9]: https://github.com/PhilanthropyDataCommons/deploy/blob/main/.github/workflows/update-service-image.yml
+[10]: https://github.com/PhilanthropyDataCommons/deploy/blob/e0d24b8a19ee563938f8211bca0485a97ff1a6e9/.github/workflows/send-tag-to-machine.yml#L22
```
Picking a block of these links at random to highlight an issue throughout these docs...

It's really cool you can do links like this! I didn't know that was possible in Markdown.

However, this style of linking makes reviewing and editing much more difficult. Commenting on a line referencing a link omits the link, making the discussion more laborious to gather context for, or, commenting on a link does not include the anchor referencing it. Likewise, editing becomes more difficult - what happens if we decide we no longer want link number 4? Do we need to renumber links 5-16? Worse, what if we want to add link number 17?

We need to be able to maintain our docs just as much as we maintain our software!

bickelj replied:
Yeah it is a pain to update all the link numbers. I'll replace them with brief descriptive text.

bickelj replied:
One of the advantages of the numbers was to prevent so many awkward line breaks and spilling-over-the-80-char mark. I can't seem to avoid that with the other style but I agree it's going to be more maintainable to keep the links in-band right at the point of reference.

jasonaowen replied:
I have found [semantic linefeeds](https://rhodesmill.org/brandon/2012/one-sentence-per-line/) to be an effective practice for formatting Markdown!

bickelj replied:
I am sure it helps maintainability but the result is really ugly when trying to read in plain text.

jasonaowen replied:
I would like to gently suggest that this is an instance where it is worthwhile to intentionally develop your taste. Maintainable code - and prose! - should be aesthetically pleasing, even if it takes some work to get there. :slightly_smiling_face:

jasonaowen reviewed (dismissed):
Thank you for working on documenting these processes, @bickelj!

I've tried to focus primarily on the content, rather than the process it describes. Sorry there's so much; I hope most of these suggestions are lightweight and uncontroversial enough to knock out quickly!

bickelj commented:
Oh, now after expanding, I see some more comments to address above.

bickelj committed:
Add `DEPLOYING.md` and `UPGRADING.md` documents

The documentation in this repo so far has descriptions of the elements
in a potential deployment pipeline but lacked a step-by-step guide to
deploy PDC in the actual instance of the PDC's pipeline. This commit
resolves the gap by adding more concrete guides in `DEPLOYING.md` and
`UPGRADING.md`. "Deploying" here means "how to deploy to production"
and "upgrading" here means "how to update a service-level dependency
such as PostgreSQL in production."

Issue https://code.librehq.com/ots/infra/admin-tasks/-/issues/63
Issue https://code.librehq.com/ots/infra/admin-tasks/-/issues/64

bickelj commented:
This should be closer.

jasonaowen reviewed (approved):
Thanks for working through all those suggestions, @bickelj! I hope you think the final result is better; I do. :smiley:

bickelj commented:
> Thanks for working through all those suggestions, @bickelj! I hope you think the final result is better; I do. smiley

No doubt it is better!
